File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.py
def solution(p):
    
    def check1(arr): #올바른 문자열 확인
        stack=[]
        for a in arr:
            if a=='(':
                stack.append(a)
            else:
                if stack:
                    stack.pop()
                else:
                    return False #올바른게
        return not stack
    
    def check2(arr):
        cnt1=0
        cnt2=0
        for a in arr:
            if a=='(':
                cnt1+=1
            else:
                cnt2+=1
        return cnt1==cnt2
    
    def divide(arr): #u와v로분리해주는 함수
        u1=''
        v1=''
        for i in range(len(arr)):
            u1+=arr[i]
            if check2(u1):
                v1=arr[i+1:]
                break
        return u1, v1
    
    def process(p):
        if p=='':
            return ''
        answer = ''
        u, v = divide(p)
        if check1(u): #u가 올바른 문자열 이라면 1단계부터 실행
            answer=answer+u+process(v)
        else:
            answer=answer+'('
            answer=answer+process(v)
            answer=answer+')'
            u=u[1:len(u)-1]
            u_reverse=''
            for i in u:
                if i=='(':
                    u_reverse=u_reverse+')'
                else:
                    u_reverse=u_reverse+'('
            answer=answer+u_reverse
        return answer
    return process(p)

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.java
import java.util.Stack;

class Solution {
    public String solution(String p) {
        return dfs(p);
    }

    private String dfs(String p){
        //균형잡힌 문자열로 분리
        if(p.isEmpty()){
            return "";
        }
        String u_str="", v_str="";
        int index = divide(p);

        u_str = p.substring(0, index+1);
        v_str = p.substring(index+1);

        if(check(u_str)) { //u_str가 올바른 문자열이라면
            return u_str + dfs(v_str); //u_str은 이미 올바르니까 v_str만 제대로 정렬
        }

        else{
            String new_u_str = "(" + dfs(v_str) + ")";

            u_str = u_str.substring(1, u_str.length()-1); //u의 첫번째와 마지막 문자 제거

            //문자 뒤집기
            for(int i=0; i<u_str.length(); i++){
                new_u_str += (u_str.charAt(i) == '(') ? ')' : '(';
            }
            return new_u_str;
        }
    }

    //균형잡힌 문자열로 분리
    private int divide(String p){
        int left=0, right=0;
        for(int i=0; i<p.length(); i++){
            if (p.charAt(i)=='('){
                left++;
            }
            else{
                right++;
            }
            if(left!=0 && right!=0 && left==right) {
                return i;
            }
        }
        return 0;
    }

    private boolean check(String p){ //올바른 문자열인지 확인하는 함수
        Stack<Character> stack = new Stack<>();
        for(int i=0; i<p.length(); i++){
            if (p.charAt(i)=='('){
                stack.push('(');
            }
            else {
                if(stack.isEmpty()){
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
}
